File: models/individual.py
from models.genes import GenesAs8BitArray

class Individual:
    def __init__(self, config, birth_year):
        self.config = config
        self.birth_year = birth_year
        self.genes = GenesAs8BitArray(config)

    def age(self, current_year):
        return current_year - self.birth_year

    def isReadyToDie(self, current_year):
        return self.age(current_year) >= self.config.max_age

    def canBreed(self, current_year):
        age = self.age(current_year)
        config = self.config # optimize
        return (age >= config.min_breeding_age and age <= config.max_breeding_age)

    def inheritFromParentsAndMutate(self, father, mother):
        self.genes.inheritFrom(father.genes, mother.genes)
        self.genes.mutate()

File: src/uk/edu/populationfitness/models/Individual.java
package uk.edu.populationfitness.models;

/**
 * Created by pete.callaghan on 03/07/2017.
 */
public class Individual {
    private Config config;

    private int birth_year;

    public Genes genes;

    public double fitness;

    public Individual(Config config, int birth_year){
        this.config = config;
        this.birth_year = birth_year;
        this.genes = new Genes(config);
    }

    public int age(int current_year){
        return current_year - birth_year;
    }

    public boolean isReadyToDie(int current_year){
        return age(current_year) >= config.max_age;
    }

    public boolean canBreed(int current_year){
        int age = age(current_year);
        return age >= config.min_breeding_age && age <= config.max_breeding_age;
    }

    public void inheritFromParentsAndMutate(Individual mother, Individual father){
        genes.inheritFrom(mother.genes, father.genes);
        genes.mutate();
    }
}
